xdsl/ir/affine/affine_map.py

```diff
@@ -251,6 +251,40 @@ def compress_dims(self, selectors: Sequence[bool]) -> AffineMap:
             new_dims, new_symbols, result_num_dims, self.num_symbols
         )
 
+    def used_dims(self) -> set[int]:
+        """
+        Return all dimensions used in the map as a set
+
+        Example:
+        ```
+        (d0, d1) -> (d0) gives {d0}
+        (d0, d1, d2) -> (d0, d2) gives {d0, d2}
+        ```
+        """
+        result: set[int] = set()
+
+        for expr in self.results:
+            result = result.union(expr.used_dims())
+
+        return result
```

Comment:
Set construction feels expensive enough to me to only want to do it once:
```suggestion
        return {expr.position for res_expr in self.results for expr in res_expr.dfs() if isinstance(expr, AffineDimExpr)}
```